pkg/multiClusterWatch/main.go

```diff
@@ -0,0 +1,165 @@
+	if w.watchers == nil {
+		w.watchers = map[string]Watcher{}
+	}
+
+	if w.watchers[config.Host] != nil {
+		return w.watchers[config.Host], nil
+	}
+
+	watcher, err := NewClusterWatcher(w.Manager, config, w.HandlerFactory)
+	if err != nil {
+		return nil, err
+	}
+
+	w.watchers[config.Host] = watcher
+	return watcher, nil
+}
+
+func (w *ClusterWatcher) Start(ctx context.Context) error {
+	log.Log.Info("Starting cluster watcher", "name", w.ClusterName)
+
+	ingresses, err := w.client.NetworkingV1().Ingresses("").List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, i := range ingresses.Items {
+		log.Log.Info("New cluster client, can see ingress", "name", i.Namespace+"/"+i.Name, "host", w.ClusterName)
+	}
+
+	informerFactory := informers.NewSharedInformerFactory(w.client, RESYNC_PERIOD)
+
+	informer := informerFactory.Networking().V1().Ingresses().Informer()
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			log.Log.Info("got add event for ingress", "cluster watcher", w.ClusterName, "ingress", obj.(*networkingv1.Ingress).Namespace+"/"+obj.(*networkingv1.Ingress).Name)
+			current := obj.(*networkingv1.Ingress)
+			target := current.DeepCopy()
+			targetAccessor := traffic.NewIngress(target)
+			w.Handler.Handle(ctx, targetAccessor)
+			//todo handle requeue and errors
+			if !equality.Semantic.DeepEqual(current, target) {
+				//write back to cluster
+				w.client.NetworkingV1().Ingresses(target.Namespace).Update(ctx, target, metav1.UpdateOptions{})
+			}
+		},
+		UpdateFunc: func(old, obj interface{}) {
+			log.Log.Info("got update event for ingress", "cluster watcher", w.ClusterName, "ingress", obj.(*networkingv1.Ingress).Namespace+"/"+obj.(*networkingv1.Ingress).Name)
+			current := obj.(*networkingv1.Ingress)
+			target := current.DeepCopy()
+			targetAccessor := traffic.NewIngress(target)
+			w.Handler.Handle(ctx, targetAccessor)
+			//todo handle requeue and errors
+			if !equality.Semantic.DeepEqual(current, target) {
+				//write back to cluster
+				w.client.NetworkingV1().Ingresses(target.Namespace).Update(ctx, target, metav1.UpdateOptions{})
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			log.Log.Info("got delete event for ingress", "cluster watcher", w.ClusterName, "ingress", obj.(*networkingv1.Ingress).Namespace+"/"+obj.(*networkingv1.Ingress).Name)
+			current := obj.(*networkingv1.Ingress)
+			target := current.DeepCopy()
+			targetAccessor := traffic.NewIngress(target)
+			w.Handler.Handle(ctx, targetAccessor)
+			//todo handle requeue and errors
+			if !equality.Semantic.DeepEqual(current, target) {
+				//write back to cluster
+				w.client.NetworkingV1().Ingresses(target.Namespace).Update(ctx, target, metav1.UpdateOptions{})
+			}
+		},
+	})
+
+	informerFactory.Start(ctx.Done())
+	informerFactory.WaitForCacheSync(ctx.Done())
+
+	log.Log.Info("started watcher events", "cluster watcher", w.ClusterName)
+
+	<-ctx.Done()
```

Comment:
I am guessing start is called in it own go routine

Reply:
have we tried with multiple secrets yet?

Reply:
Not yet. This does need testing, but yes, the kubebuilder manager runs this in a separate goroute.